integration-tests/tests/src/tests/sync/app.ts

```diff
@@ -0,0 +1,322 @@
+    });
+
+    it("from config", function (this: ConfigContext) {
+      const app = new Realm.App(this.conf);
+      expect(app).instanceOf(Realm.App);
+    });
+
+    it("from string", function (this: ConfigContext) {
+      const app = new Realm.App(this.conf.id);
+      expect(app).instanceOf(Realm.App);
+      expect(app.id).equals(this.conf.id);
+    });
+
+    it("throws on undefined app", function () {
+      //@ts-expect-error creating an app without a config should fail
+      expect(() => new Realm.App()).throws(Error, "Invalid arguments: 1 expected, but 0 supplied.");
+    });
+
+    it("throws on invalid input", function () {
+      //@ts-expect-error creating an app with an invalid config should fail
+      expect(() => new Realm.App(1234)).throws(Error, "Expected either a configuration object or an app id string.");
+    });
+
+    it("throws on invalid baseURL", function () {
+      const invalid_url_conf = { id: "smurf", baseUrl: "http://localhost:9999" };
+      const app = new Realm.App(invalid_url_conf);
+
+      const credentials = Realm.Credentials.anonymous();
+      console.log(credentials);
+      console.log(app.logIn);
+      return new Promise<void>((resolve, reject) => {
+        return app
+          .logIn(credentials)
+          .then((user) => {
+            return reject(`Able to log in with config ${JSON.stringify(invalid_url_conf)}`);
+          })
+          .catch((err) => {
+            expect(err.message).contains(
+              "request to http://localhost:9999/api/client/v2.0/app/smurf/location failed, reason: connect ECONNREFUSED",
+            );
+            return resolve();
+          });
+      });
+    });
+
+    it("throws on non existing app", async function () {
+      const app = new Realm.App(this.conf);
+      const credentials = Realm.Credentials.anonymous();
+      let failed = false;
+      await app.logIn(credentials).catch((err) => {
+        failed = true;
+        expect(err.message).equals("cannot find app using Client App ID 'smurf'");
+      });
+      expect(failed).to.be.true;
+    });
+  });
+
+  describe("with valid app", async () => {
+    importAppBefore("with-db");
+
+    it("logins successfully ", async function (this: Mocha.Context & AppContext & RealmContext) {
+      expect(this.app).instanceOf(Realm.App);
+      const credentials = Realm.Credentials.anonymous();
+      const user = await this.app.logIn(credentials);
+      expect(user).instanceOf(Realm.User);
+      expect(user.deviceId).to.not.be.null;
+      expect(user.providerType).equals("anon-user");
+      await user.logOut();
+    });
+
+    it("throws on login with non existing user ", async function (this: Mocha.Context & AppContext & RealmContext) {
+      expect(this.app).instanceOf(Realm.App);
+      const credentials = Realm.Credentials.emailPassword("me", "secret");
+      let didFail = false;
+      const user = await this.app.logIn(credentials).catch((err) => {
```

Comment:
This could be refactored to use `expect().to.throw()`

Reply:
That only works for synchronously thrown errors - we should use `expect().should.be.rejectedWith()` to assert rejection of promises. That being said, I don't think that'll be able to replace this code entirely as it's not simply matching on the error message, but also asserts the `code` to be `50` 🤔